src/django_components/extension.py

```diff
@@ -0,0 +1,580 @@
+from functools import wraps
+from typing import TYPE_CHECKING, Any, Callable, Dict, List, NamedTuple, Tuple, Type, TypeVar
+
+from django.template import Context
+
+from django_components.app_settings import app_settings
+from django_components.util.misc import snake_to_pascal
+
+if TYPE_CHECKING:
+    from django_components import Component
+    from django_components.component_registry import ComponentRegistry
+
+
+TCallable = TypeVar("TCallable", bound=Callable)
+
+
+################################################
+# HOOK TYPES
+#
+# This is the source of truth for what data is available in each hook.
+# NOTE: These types are also used in docs generation, see `docs/scripts/reference.py`.
+################################################
+
+
+class OnComponentClassCreatedContext(NamedTuple):
```

Comment:
Within the hooks, various data can be accessed on the given "Context" object.

All these data classes are also used for generating from the docstrings on these properties.